src/simulators.rs

```diff
@@ -1,61 +1,192 @@
+use std::collections::VecDeque;
 use generators::Generator;
 
 // Packet holds the value of the time in ticks that it was generated at.
+#[derive(Debug)]
 pub struct Packet(pub u32);
 
-// Client generates packets according as per the provided generators::Generator.
+#[derive(Debug)]
+pub struct ClientStatistics {
+    pub packets_generated: u32,
+}
+
+impl ClientStatistics {
+    fn new() -> ClientStatistics {
+        ClientStatistics { packets_generated: 0 }
+    }
+}
+// Client generates packets according as per the parametrized generators::Generator.
 #[allow(dead_code)]
-struct Client<G: Generator> {
-    // Next time value that a tick will be generated
-    next_gen: u32,
-    // The "length" of time of one tick.
+pub struct Client<G: Generator> {
+    // TODO(irfansharif): Update comments.
     resolution: f64,
+    ticker: u32,
     generator: G,
+    pub statistics: ClientStatistics,
 }
 
+// TODO(irfansharif): Add comments.
 impl<G: Generator> Client<G> {
     #[allow(dead_code)]
-    fn new(generator: G, resolution: f64) -> Client<G> {
+    pub fn new(generator: G, resolution: f64) -> Client<G> {
         Client {
-            next_gen: 0,
+            ticker: generator.next_event(resolution),
             generator: generator,
+            statistics: ClientStatistics::new(),
+            resolution: resolution,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn tick(&mut self) -> bool {
+        // TODO(irfansharif): Resolution mismatch; no possibility of generating multiple packets.
+        if self.ticker == 0 {
+            self.statistics.packets_generated += 1;
```

Comment:
Maybe we can express this as 
```rust
self.ticker -= 1
if self.ticker <= 0 { 
   // Code
   true
} else {
   false
}
```
, since this would accomplish the same logic and get rid of this code duplication. It looks like since we are now initializing client to generate a packet at a random time rather than always at 0, the only case where ticker would be 0 at the beginning of a call to ticker would be if the generator ended up generating a 0 due to truncation, in which case we would see a value of -1, which *could* be okay...

Reply:
This is more of a nit than anything though

Reply:
yea the reason was to avoid getting to -1, originally had it that way before getting buffer underflows (ticker is a u32, as I think it should be).